numpy/linalg/linalg.py

```diff
@@ -651,17 +651,19 @@ def matrix_power(a, n):
         return fmatmul(a, a)
 
     elif n == 3:
-        return fmatmul(fmatmul(a, a), a)
+        # create and use buffered space
+        buffer = fmatmul(a, a)
+        return fmatmul(buffer, a, out=buffer)
```

Comment:
Seems clear that this will help.

Reply:
Do we even support in-place matrix multiply? (Does BLAS even support it?)

I honestly expect the `matmul` ufunc (and `np.dot`) will just run overlap detection and create an *additional* internal buffer here.  (I am having problems with `timeit` not running my setup code right now).

Unless there are some clear timings, We should probably keep the new tests and maybe see how to improve the `n>3` case, where I expect something might work (if it is worth it).  Otherwise, first we need to dig into avoiding the additional copy in `np.dot` and `np.matmul` first (and making sure that is correct).